Some bugfixes

Rob23oba opened the pull request:
Nothing as big this time, just some bugfixes, worlds and a level enter stack.

L4Vo5 commented:
Ehh, I don't really wanna add worlds. I feel like they add a lot of conceptual complexity with no real benefit other than parity with the base game. Can you remove that?

Rob23oba commented:
Fair, but I'd think hubs are an important part of level packs and making the distinction of worlds and levels for the sake of exiting levels would at least make sense - it doesn't need to be a full implementation of the mechanic. For now, I've reverted the changes though.

L4Vo5 commented:
The thing is, unlike in the original game, you can put entries in any level. So it probably makes more sense that pressing backspace or completing a level would just take you back to the last level you were in, regardless of what level it is, instead of using a whole world system.
So you can still make hubs and they'll work as expected, the only difference being that you can't "complete" them after finishing X amount of levels. But is that kind of progression really so important a feature for a fan-made level pack? I was planning to reuse stone keys for a new mechanic instead. And, if you _really_ want it, you can probably still implement world completion with some amount of salvaging nonsense.

Rob23oba commented:
I think the most important distinctions of a world are that first of all they don't have a goal. This behavior is currently impossible (sure, it is not necessary for worlds to be unbeatable in that sense but it feels like it'd make sense).
Second of all, you can't exit out of a world, instead you use world entries to go from one place to another. This behavior is also technically unneeded, but given exiting a level would be implemented, it would also make sense that you could do so recursively, in which case you'd need an enter stack. But if you have multiple hubs that don't just have a tree-like structure but also have loops (e.g. shortcuts like FP -> Main Hub -> FP), you are going to end up with a large stack that you can't get rid of without going back through all worlds in between. I understand that these are many variables which have to line up to have for that to be a problem but yeah.
These two problems *could* be solved using the notion of a world that has no goals and you can't exit out of. You could also split them up into two options but I feel like that'd be a bit too much.
Anyways, I don't really care much about the whole "world completion" thing. Really, it's just something I implemented because it wasn't too hard to do. Also I don't think this discussion is too important right now - there's still saving and other stuff like that - which I also have some potentially controversial ideas about - let's see.

Rob23oba commented:
... here it goes

L4Vo5 commented:
Alright, should probably take a proper look now, before the changes I was making start to conflict more heavily

L4Vo5 commented:
... ok yeah that saving stuff is _very_ controversial. any chance it could be a different PR? I'd like to prioritize the other stuff so I can hopefully merge it sooner rather than later

Rob23oba commented:
I've put the new save load in a separate branch now.

L4Vo5 commented:
ah, so that's how "reviews" work. earlier it kinda looked like the progress of it was public before submitting it

Rob23oba commented on level_elements/level.gd:
```diff
@@ -127,6 +127,13 @@ func _ready() -> void:
+	var camera_pos2 = camera_pos + camera.get_viewport_rect().size
+	if mouse_pos.x < camera_pos.x or mouse_pos.y < camera_pos.y \
+		or mouse_pos.x >= camera_pos2.x or mouse_pos.y >= camera_pos2.y:
+		hover_highlight.stop_adapting()
```
If you have something on the edge of the level and move the mouseover out of it, the mouseover would previously stay (at least after my mouseover change)

Rob23oba commented on level_elements/level.gd:
```diff
@@ -233,66 +238,70 @@ func try_open_door(door: Door) -> void:
-signal key_gui_input(event: InputEvent, key: Key)
-signal entry_gui_input(event: InputEvent, entry: Entry)
-signal salvage_point_gui_input(event: InputEvent, salvage_point: SalvagePoint)
+signal element_gui_input(event: InputEvent, door: Door, type: Enums.level_element_types)
```
(small oversight)

Rob23oba commented on level_elements/level.gd:
```diff
@@ -362,21 +380,12 @@ func place_goal(coord: Vector2i) -> void:
 	if level_data.tiles.get(tile_coord): return
 	if is_space_occupied(Rect2i(tile_coord * 32, Vector2i(32, 32)), [&"tiles"]): return
-	level_data.tiles[tile_coord] = 1
+	level_data.tiles[tile_coord] = true
```
Okay, then I'd do so everywhere else, I think it doesn't do that yet in the save load. Also in this case maybe use `level_data.tiles.has(tile_coord)` instead of `get(tile_coord)` to make sure it doesn't unnecessarily need to query the value and cast it

Rob23oba commented on data/level_pack_state_data.gd:
```diff
@@ -66,32 +68,115 @@ func _on_deleted_level(level_id: int) -> void:
+	data.store_uint(pack_state_meta_category.completed_levels)
+	data.store_uint(completed_levels.size())
+	for b in completed_levels:
+		data.store_8(b)
```
(uh wait where does it say the completed_levels.size() is saved in a u8)

Rob23oba commented on misc/saving_versions/save_load_v5.gd:
```diff
@@ -0,0 +1,454 @@
+		while min > buffer.get_size():
+			buffer.resize(buffer.get_size() * 2)
+	
+	func store_8(value: int) -> void:
```
they work for both signed and unsigned due to how 2's complement works

Rob23oba commented on data/door_data.gd:
```diff
@@ -96,7 +96,9 @@ func duplicated() -> DoorData:
 #	assert(PerfManager.start("DoorData::duplicated (locks)"))
 	for l in locks:
-		dupe.locks.push_back(l.duplicated())
+		# "Please don't push_back on these or anything!! use add_lock or remove_lock"
```
there is also the lock.changed.connect(emit_changed)..?

Rob23oba commented on misc/saving_versions/save_load_v5.gd:
```diff
@@ -0,0 +1,454 @@
+
+class ByteWriter:
+	extends RefCounted
+	var buffer: StreamPeerBuffer
```
> feels weird to use StreamPeerBuffer for entirely the wrong purpose
I think this is what StreamPeerBuffer actually is for lol. After some simple performance tests, it turns out StreamPeerBuffer has ~20% better reading/writing times as PackedByteArray

Rob23oba commented on misc/saving_versions/save_load_v5.gd:
```diff
@@ -0,0 +1,452 @@
+	
+	## Unsigned LEB128
+	## Special case: 0x81 0x00 -> -1 for invalid entries / SIDs
+	func store_uint(value: int) -> void:
```
hey, they are not *that* complicated
